ngramoffset.go

```diff
@@ -0,0 +1,346 @@
+// Copyright 2021 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zoekt
+
+import (
+	"sort"
+)
+
+// simple interface type check for implementations
+var (
+	_ ngramOffset = &arrayNgramOffset{}
+	_ ngramOffset = &asciiNgramOffset{}
+	_ ngramOffset = combinedNgramOffset{}
+)
+
+// shrinkUint32Slice copies slices with excess capacity to precisely seized ones
+// to avoid wasting memory. It should be used on slices with long static durations.
+func shrinkUint32Slice(a []uint32) []uint32 {
+	if cap(a)-len(a) < 32 {
+		return a
+	}
+	out := make([]uint32, len(a))
+	copy(out, a)
+	return out
+}
+
+type ngramOffset interface {
+	Get(ngram) simpleSection
+	DumpMap() map[ngram]simpleSection
+	SizeBytes() int
+}
+
+func makeNgramOffsetEmpty() ngramOffset {
+	return makeArrayNgramOffset(nil, nil)
+}
+
+func makeNgramOffset(ngrams []ngram, offsets []uint32) ngramOffset {
+	return makeCombinedNgramOffset(ngrams, offsets)
+}
+
+type topOffset struct {
+	top uint32
+	off int
```

Comment:
I think this should be a uint32? If not, you need to update your size calculation.